### PR TITLE
Add scalar function CURRENT_TIME returning current time UTC with microsecond precision

### PR DESCRIPTION
**Pull request by marregui**
As per https://github.com/crate/crate/issues/9662 

scalar function that returns the new type `time with time zone`.

### CODE REVIEW

**Comment by marregui** [comment]
I have `TransactionContext.currentTimeMillis` to workout micro precision, which I could not do.
I played with the implementations of TransactionContext, and found that I could get rid of DateTimeUtils.currentTimeMillis() and use instead:

```
Instant i = Clock.systemUTC().instant();
long currentTimeMicro = Math.addExact(
    Math.multiplyExact(i.toEpochMilli(), 1000L),  
    Math.floorDiv(i.getNano() / 1000L));
```
That would involve updating some tests because they set the value to return.

For simplicity I would add a new method to `TransactionContext.currentTimeMicros`.

File: docs/general/ddl/data-types.rst [hunk 1]
```diff
@@ -437,6 +443,69 @@ In these expressions, the desired time zone is specified as a string
 The scalar function :ref:`TIMEZONE <scalar-timezone>` (zone, timestamp) is
 equivalent to the SQL-conforming construct timestamp AT TIME ZONE zone.
 
+.. _time-data-type:
+
+time with time zone
+-------------------
+
+The time type consist of time followed by an optional time zone.
```
Comment (infoverload) [inline docs/general/ddl/data-types.rst:451]: consist = consists

File: docs/general/ddl/data-types.rst [hunk 2]
```diff
@@ -437,6 +443,69 @@ In these expressions, the desired time zone is specified as a string
 The scalar function :ref:`TIMEZONE <scalar-timezone>` (zone, timestamp) is
 equivalent to the SQL-conforming construct timestamp AT TIME ZONE zone.
 
+.. _time-data-type:
+
+time with time zone
```
Comment (infoverload) [inline docs/general/ddl/data-types.rst:448]: Time with time zone

or

Time with Time Zone

?

Comment (marregui) [reply docs/general/ddl/data-types.rst:448]: in this case I would keep lowercase, as it is how you normally would spell it on the crate shell.

**Comment by mfussenegger** [comment]
Could you rebase this now? Otherwise it's hard to review

**Comment by marregui** [comment]
retest this please

the job seems to have been cancelled

File: docs/general/builtins/scalar.rst [hunk 3]
```diff
@@ -833,9 +850,9 @@ are given.
 
 .. NOTE::
 
-   The ``CURRENT_TIMESTAMP`` will be evaluated  using javas
-   ``System.currentTimeMillis()``. So its actual result depends on the
-   underlying operating system.
+   The ``CURRENT_TIMESTAMP`` and ``CURRENT_TIME`` expressions will be evaluated
+   using JMV's ``Clock.systemUTC().instant()``. So its actual result depends on
+   the underlying operating system, hardware and the JVM implementation.
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst:855]: Maybe we could omit the implementation details here and just state that the time returned depends on the system clock and JVM implementation.

File: docs/general/builtins/scalar.rst [hunk 4]
```diff
@@ -805,6 +805,23 @@ The following fields are supported:
...
 .. _`Joda-Time`: http://www.joda.org/joda-time/
 
+.. _current_time:
+
+``CURRENT_TIME``
+----------------
+
+The ``CURRENT_TIME`` expression returns the time in microseconds
+since midnight UTC at the time the SQL statement was handled. Therefore,
+the same time value is returned for every invocation of a single statement.
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst:815]:
```suggestion
the same time value is returned for every invocation within a single statement.
```

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -459,13 +460,20 @@ protected Symbol visitExpression(Expression node, ExpressionAnalysisContext cont
...
+            CurrentTime.Type type = node.getType();
+            if (!(type.equals(CurrentTime.Type.TIMESTAMP) || type.equals(CurrentTime.Type.TIME))) {
                 visitExpression(node, context);
             }
+            String funcName = CurrentTimestampFunction.NAME;
+            int defaultPrecision = CurrentTimestampFunction.DEFAULT_PRECISION;
+            if (type.equals(CurrentTime.Type.TIME)) {
+                funcName = CurrentTimeFunction.NAME;
+                defaultPrecision = CurrentTimeFunction.DEFAULT_PRECISION;
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:472]: I think a single conditional would make this a bit easier to parse. Maybe with something like this?

```java
String funcName;
int defaultPrecision;
switch (type) {
    case CurrentTime.Type.TIMESTAMP:
        funcName = ...
        break;

    case CurrentTime.Type.TIME:
        funcName = ...
        break;


    default:
      return visitExpression(node, context);
}
return allocateFunction(...
```


I think it would also be an option to make the default precision argument optional and keep the "what value is the default" knowledge encapsulated within the function itself. So that the expression analyzer doesn't add the argument if it is not specified.

File: benchmarks/src/test/java/io/crate/execution/TimePrecisionIncreaseTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        @Override
+        public long currentTimeMicros() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public SessionSettings sessionSettings() {
+            throw new UnsupportedOperationException();
+        }
+    }
```
Comment (mfussenegger) [inline benchmarks/src/test/java/io/crate/execution/TimePrecisionIncreaseTest.java:71]: Not sure if the extension of the TransactionContext is valuable here. A abstract class definition with the `currentTimeMicros` would be sufficient for the benchmark. Or even just define the two different methods directly without any classes at all

File: server/src/main/java/io/crate/metadata/TransactionContext.java [hunk 7]
```diff
@@ -23,38 +23,77 @@
...
-    static TransactionContext of(SessionSettings sessionSettings) {
-        return new StaticTransactionContext(sessionSettings);
+public class TransactionContext {
+
+    public static TransactionContext of(SessionSettings sessionSettings) {
+        return new TransactionContext(sessionSettings);
     }
 
-    long currentTimeMillis();
+    private static final AtomicReference<Supplier<Long>> MILLIS_PROVIDER = new AtomicReference<>(null);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/TransactionContext.java:38]: I think it would be cool if we could de-couple more of the time related logic into more specific components. The TransactionContext should only be concerned about caching the value retrieved to ensure the same value is re-used across the transaction. It shouldn't know too much about how that value is produced.

For example we could introduce a `Clocks` class that encapsulates the logic about fixing/mocking the current time.
Maybe even work with the `Clock` classes directly instead of the `Supplier<Long>`.

There is both

```
Clock.systemUTC();
Clock.fixed(..);
```

File: docs/general/builtins/scalar.rst [hunk 8]
```diff
@@ -805,6 +805,23 @@ The following fields are supported:
 .. _`available time zones`: http://www.joda.org/joda-time/timezones.html
 .. _`Joda-Time`: http://www.joda.org/joda-time/
 
+.. _current_time:
+
+``CURRENT_TIME``
+----------------
+
+The ``CURRENT_TIME`` expression returns the time in microseconds
+since midnight UTC at the time the SQL statement was handled. Therefore,
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst:814]: Btw:

```
 * The clock implementation provided here is based on the same underlying clock
 * as {@link System#currentTimeMillis()}, but may have a precision finer than
 * milliseconds if available.
 * However, little to no guarantee is provided about the accuracy of the
 * underlying clock. Applications requiring a more accurate clock must implement
 * this abstract class themselves using a different external clock, such as an
 * NTP server.
```

I think we've to mention this in some way

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(SIGNATURE, (signature, args) -> new CurrentTimeFunction());
+    }
+
+    @Override
+    @SafeVarargs
+    public final TimeTZ evaluate(TransactionContext txnCtx, Input<Integer>... args) {
+        Integer precision = DEFAULT_PRECISION;
+        if (args.length == 1) {
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:67]: the function always has[ one argument](https://github.com/crate/crate/pull/10061/files#diff-87f484c9b83a7bd4e2f622b44c4911ddR47), so this check is probably obsolete?

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            precision = args[0].value();
+            if (precision == null) {
+                throw new IllegalArgumentException(String.format(
+                    Locale.ENGLISH, "NULL precision not supported for %s", NAME));
+            }
+        }
+        if (precision < 0 || precision > DEFAULT_PRECISION) {
```
Comment (kovrus) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:74]: minor: imho, we can probably fail early in the analyzer in the case when the precision is provided and it is wrong. Otherwise, we allocate a function call with the already known wrong precision value. Also, if the function is called without precision, this condition is simply unnecessary.

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:74]: I don't think we should micro optimize for error cases like that. I think it is more important to encapsulate / narrow the responsibility. The ExpressionAnalyzer should be concerned with converting the parser expression to symbols and not know too much details about internals of functions and so on.

So I'd be in favor of keeping this logic here inside the function (see also my other comment about also moving the knowledge about the default value)

Comment (kovrus) [reply server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:74]: right, makes sense, otherwise, it would scatter the function's logic across analyzer and function implementation.

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java (hunk 9 above)
Comment (marregui) [reply server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:67]: I have corrected this, the function can have no arguments indeed, so I am registering two signatures, one with and integer arg, the other with no args, that way defaults are defined in the function implementing classes, and the number of args needs to be checked.

File: docs/general/builtins/scalar.rst [hunk 11]
```diff
@@ -824,6 +824,12 @@ synopsis::
 the time will have second precision, no fractional seconds (microseconds)
 are given.
 
+.. NOTE::
+   No guarantee is provided about the accuracy of the underlying clock.
+   Applications requiring a more accurate clock must implement abstract
+   class ``java.time.Clock`` themselves using a different external clock,
+   such as an NTP server.
+
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst]: This is user facing documentation and there is no way a user is going to implement a custom `Clock` class. 

I meant that we need to point out that depending on the system the accuracy of the results may be limited to milliseconds.

File: server/src/main/java/io/crate/metadata/SystemClock.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SystemClock {
+
+    private static final Clock DEFAULT_CLOCK = Clock.systemUTC();
+
+    private static final AtomicReference<Clock> SYSTEM_CLOCK = new AtomicReference<>(DEFAULT_CLOCK);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/SystemClock.java:34]:
```suggestion
    private static final AtomicReference<Clock> CURRENT_CLOCK = new AtomicReference<>(Clock.systemUTC());
```

It's not necessarily pointing to the system clock.

File: server/src/main/java/io/crate/metadata/SystemClock.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+package io.crate.metadata;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SystemClock {
+
+    private static final Clock DEFAULT_CLOCK = Clock.systemUTC();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/SystemClock.java:32]:
```suggestion
    private static final Clock SYTEM_CLOCK = Clock.systemUTC();
```

or 

```suggestion
```

and use `Clock.systemUTC` directly, it's already a singleton - so there is no gain by the assignment to a different constant.

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 14]
```diff
@@ -460,20 +460,23 @@ protected Symbol visitExpression(Expression node, ExpressionAnalysisContext cont
...
+            switch (node.getType()) {
+                case TIMESTAMP:
+                    break;
+
+                case TIME:
+                    funcName = CurrentTimeFunction.NAME;
+                    break;
+
+                default:
+                    visitExpression(node, context);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:473]: btw. was this behavior to still fall-through to the remaining code even correct before? Looks like the code outside the switch would fail if the `default` case was hit.

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            throw new IllegalArgumentException(String.format(
+                Locale.ENGLISH, "precision must range between [0..%d]", DEFAULT_PRECISION));
+        }
+        long currentTimeMicros = txnCtx.currentTimeMicros();
+        long dateDeltaMillis = Instant.ofEpochMilli(currentTimeMicros / 1000L).truncatedTo(ChronoUnit.DAYS).toEpochMilli();
+        long microsFromUTC = currentTimeMicros - (dateDeltaMillis * 1000L);
+        long f = (long) Math.pow(10, 6 - precision);
+        microsFromUTC = LongMath.divide(microsFromUTC, f, RoundingMode.DOWN) * f;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:97]: Is this the same result as `Clock.systemUTC().instant().get(ChronoField.MICRO_OF_DAY);` would return? 

If so, maybe the `TransactionContext` could be changed to cache the `instant` instead (afaik it's a singleton, so would be safe).

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 14 above)
Comment (marregui) [reply server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:473]: The missing case is ``DATE`` for scalar function ``current_date``, which at present is unsupported:

```
cr> select current_date;
SQLActionException[UnsupportedFeatureException: Unsupported expression current_date]
```

so the default will fail launch, throwing an exception:

```
 @Override
        protected Symbol visitExpression(Expression node, ExpressionAnalysisContext context) {
            throw new UnsupportedOperationException(String.format(Locale.ENGLISH,
                "Unsupported expression %s", ExpressionFormatter.formatStandaloneExpression(node)));
        }
```

File: docs/general/builtins/scalar.rst [hunk 16]
```diff
@@ -805,6 +805,29 @@ The following fields are supported:
 .. _`available time zones`: http://www.joda.org/joda-time/timezones.html
 .. _`Joda-Time`: http://www.joda.org/joda-time/
 
+.. _current_time:
+
+``CURRENT_TIME``
+----------------
+
+The ``CURRENT_TIME`` expression returns the time in microseconds
+since midnight UTC at the time the SQL statement was handled. Therefore,
```
Comment (mfussenegger) [inline docs/general/builtins/scalar.rst:814]: Reading this again I think the `Therefore, ...` implies that the prior sentence explains the behavior of `the same time value is returned for every invocation within a single
statement.` but there is no logical correlation between the two.

The first statement would actually imply that there could be different values.

I think this should be rephrased a bit to instead point out that the clock time is looked up at most once within the scope of a single query to ensure multiple `current_time` occurences evaluate to the same value.

File: server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.function.BiFunction;
+
+abstract class TimeFunction<R> extends Scalar<R, Integer> {
+
+    private final Signature signature;
+    private final FunctionInfo info;
+    private final int defaultPrecision;
+    private final BiFunction<Instant, Integer, R> logic;
+
+    TimeFunction(Signature signature, int defaultPrecision, BiFunction<Instant, Integer, R> logic) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java:46]: It's kinda hard to understand what `Integer, Integer` here are referring to in `logic` 

Why the big diverge from the prior solution?

File: server/src/main/java/io/crate/metadata/SystemClock.java [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        return withPrecision(microsFromMidnightUTC(i), precision, MICRO_PRECISION);
+    }
+
+    private static long toMicros(Instant i) {
+        return (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000L;
+    }
+
+    private static long toMillis(Instant i) {
+        return (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000_000L;
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/SystemClock.java:81]: Is this broad API surface really necessary? The prior version had a narrower surface, why was the extension necessary?

File: server/src/main/java/io/crate/metadata/TransactionContext.java [hunk 19]
```diff
@@ -23,38 +23,31 @@
...
 
-public interface TransactionContext {
+import java.time.Instant;
 
-    static TransactionContext of(SessionSettings sessionSettings) {
-        return new StaticTransactionContext(sessionSettings);
-    }
-
-    long currentTimeMillis();
+public class TransactionContext {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/TransactionContext.java:29]: Could we keep this as an itnerface?

Due to the change to caching the `Instant` I think the remaining changes to share more logic are no longer necessary.

That would avoid the `super(null);` in the `CoordinatorTxnCtx`  which seems like a hack?

File: server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java (hunk 17 above)
Comment (marregui) [reply server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java:46]: There are three related functions:

File: server/src/main/java/io/crate/metadata/SystemClock.java (hunk 18 above)
Comment (marregui) [reply server/src/main/java/io/crate/metadata/SystemClock.java:81]: My line of thinking is that all the logic in this class, now, is related to obtaining current time in either micros/millis precision, and then with user defined (additional) precision. For the case of `current_time` the date needs to be removed, leaving the API:

- `public static void setClock(Clock newClock)`: your own clock, maybe you have the need to buy an actual atomic clock and you want to hook it up. 
- `public static void setCurrentMillisFixedUTC(long millis)`: so that you control what is returned, for tests.
- `public static void setCurrentMillisSystemUTC()`: to restore the default system clock.
- `public static Instant currentInstant()`:  This is what the transaction context will end up caching, and is used to produce millis/micros.
- `public static long currentTimeMillis()`:  current millis now, no cached value of instant.
- `public static long currentTimeMillis(Instant i)`: current millis from potentially a cached value of instant.
- `public static long currentTimeMillis(Instant i, int precision)`: as above, with additional user defined precision.
- `public static long microsFromMidnightUTC(Instant i)`: current micros from potentially a cached value of instant. In the code there is no need for `public static long microsFromMidnightUTC()`, so I am not adding surface here.
- `public static long microsFromMidnightUTC(Instant i, int precision)`: as above, but with additional user defined precision.

The rest of the methods are not accessible, they are private.

All clock logic required for the time related functions is put together in this simple class.

File: server/src/main/java/io/crate/metadata/TransactionContext.java (hunk 19 above)
Comment (marregui) [reply server/src/main/java/io/crate/metadata/TransactionContext.java:29]: I agree, it looks like a hack... changing.

File: server/src/main/java/io/crate/metadata/SystemClock.java (hunk 18 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/metadata/SystemClock.java:81]: I think it is pushing too much of what the scalar functions should do into the `SystemClock` class. 
My initial comment suggesting a `Clocks` class was to have a class that is *only* providing the functionality to get a `Clock` instance or an `Instant` instance - and being able to adjust what kind of `Clock` is returned and used. Single responsibility.


The `precision` handling is specific to `current_timestamp` and `current_time` and has no generic use. So I think it doesn't belong into a generic `Clocks` or `SystemClock` class.

File: server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java (hunk 17 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java:54]: I feel like this is confusing boilerplate (creating signature *data* structures) with duplicating logic.

Yes, `NowFunction` and `CurrentTimestampFunction` could probably share the same implementation for the zero argument case - but in that case they could actually return the exact same instance, instead of introducing a common base class.


`CurrentTimeFunction` has little to do with `current_timestamp`. The signatures look similar - but that's about it.

The `TimeFunction` instances are not interchangeable (substitution principle). So the reason for this abstraction is only to "share common code" which is the signature definitions.
There are better ways than inheritance to achieve that.

File: server/src/main/java/io/crate/metadata/SystemClock.java (hunk 18 above)
Comment (marregui) [reply server/src/main/java/io/crate/metadata/SystemClock.java:81]: ok, will amend :)

**Comment by marregui** [comment]
All valid points, thank you for all the advise and the reviews!!

**Comment by marregui** [comment]
retest this please 

something bizarre went down

**Comment by marregui** [comment]
retest this please

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java [hunk 20]
```diff
@@ -21,38 +21,81 @@
...
+        if (precision < 0 || precision > MICRO_PRECISION) {
+            throw new IllegalArgumentException(String.format(
+                Locale.ENGLISH, "precision must be between [0..%d]", MICRO_PRECISION));
+        }
+
+        Instant i = txnCtx.currentInstant();
+        long dateTimeMicros = (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000L;
+        long dateMicros = i.truncatedTo(ChronoUnit.DAYS).toEpochMilli() * 1000L;
+        long f = (long) Math.pow(10, MICRO_PRECISION - precision);
+        return new TimeTZ(((dateTimeMicros - dateMicros) / f) * f, 0);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:89]: Would something like this works?

```suggestion
        Instant currentInstant = txnCtx.currentInstant();
        long microsSinceMidnight = ChronoUnit.MICROS.between(currentInstant.truncatedTo(ChronoUnit.DAYS), currentInstant);
        long factor = (long) Math.pow(10, MICRO_PRECISION - precision);
        return new TimeTZ((microsSinceMidnight / factor) * factor, 0);
```

This would make the conversion a bit more declarative and with that probably a bit easier to read.

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimestampFunction.java [hunk 21]
```diff
@@ -69,7 +75,11 @@ public final Long evaluate(TransactionContext txnCtx, Input<Integer>... args) {
                     "NULL precision not supported for %s", NAME));
             }
         }
-        return applyPrecision(millis, precision);
+        return applyPrecision(toMillis(txnCtx.currentInstant()), precision);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimestampFunction.java:78]:
```suggestion
        return applyPrecision(ChronoUnit.MILLIS.between(Instant.EPOCH, txnCtx.currentInstant()), precision);
```

File: server/src/test/java/io/crate/integrationtests/PgCatalogITest.java [hunk 22]
```diff
@@ -173,8 +173,8 @@ public void test_pg_proc_select_variadic_and_non_variadic_functions() {
             "ORDER BY proname");
         assertThat(printedTable(response.rows()), is(
             "-395638146| array_difference| -1861355723| 1000.0| 0| false| false| true| 2277| [2277, 2277]| NULL| array_difference\n" +
-            "726540318| current_timestamp| -1861355723| 0.0| 0| false| false| false| 1184| [23]| NULL| current_timestamp\n" +
             "726540318| current_timestamp| -1861355723| 0.0| 0| false| false| false| 1184| []| NULL| current_timestamp\n" +
+            "726540318| current_timestamp| -1861355723| 0.0| 0| false| false| false| 1184| [23]| NULL| current_timestamp\n" +
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/PgCatalogITest.java]: Why did this change?  Could it be that this is flaky because the `ORDER BY` is only on the name?

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java (hunk 20 above)
Comment (marregui) [reply server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:89]: oh! yes, higher abstraction is better. For me it is more readable the way I wrote it, but that is subjective. If allowed I will go functional.

File: server/src/test/java/io/crate/integrationtests/PgCatalogITest.java (hunk 22 above)
Comment (marregui) [reply server/src/test/java/io/crate/integrationtests/PgCatalogITest.java]: CurrentTimestampFunction has now two signatures being registered, to cover the case `current_timestamp` as well as `current_timestamp(3)`. It is flaky in  the sense that if I change the order of signature registration, I have to change the test. I will make it resilient, thank you for spotting this.

File: server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+        if (args.length == 1) {
+            precision = args[0].value();
+            if (precision == null) {
+                throw new IllegalArgumentException(String.format(
+                    Locale.ENGLISH,
+                    "NULL precision not supported for %s",
+                    signature.getName().name()));
+            }
+        }
+        if (precision < 0 || precision > MICRO_PRECISION) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java:80]: This could move into the previous `if` block. We know that the precision is valid if it comes from the default value if there are 0 arguments.

File: server/src/main/java/io/crate/metadata/CoordinatorTxnCtx.java [hunk 24]
```diff
@@ -37,7 +37,7 @@
 public final class CoordinatorTxnCtx implements TransactionContext {
 
     private final SessionContext sessionContext;
-    private Long currentTimeMillis = null;
+    private volatile Instant currentInstant;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/CoordinatorTxnCtx.java:40]: Before `currentTimeMillis` wasn't volatile, so it shouldn't be necessary for `currentInstant` to be volatile now.

File: server/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java [hunk 25]
```diff
@@ -1052,7 +1052,7 @@ public void testCreateTableGeneratedColumnWithCast() {
     @Test
     public void testCreateTableWithCurrentTimestampAsGeneratedColumnIsntNormalized() {
         BoundCreateTable analysis = analyze(
-            "create table foo (ts timestamp with time zone GENERATED ALWAYS as current_timestamp)");
+            "create table foo (ts timestamp with time zone GENERATED ALWAYS as current_timestamp(3))");
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/analyze/CreateAlterTableStatementAnalyzerTest.java:1055]: This change kinda invalidates the comment further below: `assertThat(generatedColumnsMapping.get("ts"), is("current_timestamp(3)")); // 3 is the default precision`. Either the change should be reverted or the comment removed.

File: server/src/test/java/io/crate/integrationtests/TransportSQLActionClassLifecycleTest.java [hunk 26]
```diff
@@ -590,7 +567,7 @@ public void selectWhereEqualCurrentTimestamp() throws Exception {
         SQLResponse response = execute("select * from sys.cluster where current_timestamp = current_timestamp");
         assertThat(response.rowCount(), is(1L));
 
-        SQLResponse newResponse = execute("select * from sys.cluster where current_timestamp > current_timestamp");
+        SQLResponse newResponse = execute("select * from sys.cluster where current_timestamp < current_timestamp");
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/TransportSQLActionClassLifecycleTest.java:570]: Why the change here? Semantics should still be the same?

File: server/src/test/java/io/crate/metadata/doc/DocIndexMetaDataTest.java [hunk 27]
```diff
@@ -1,5 +1,6 @@
 package io.crate.metadata.doc;
 
+import com.carrotsearch.randomizedtesting.annotations.Seed;
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/metadata/doc/DocIndexMetaDataTest.java:3]:
```suggestion
```

File: server/src/test/java/io/crate/integrationtests/TransportSQLActionClassLifecycleTest.java (hunk 26 above)
Comment (marregui) [reply server/src/test/java/io/crate/integrationtests/TransportSQLActionClassLifecycleTest.java:570]: yes, they are the same

**Comment by mfussenegger** [comment]
Please make sure to squash the commits to something sensible before merging

**Comment by marregui** [comment]
Thank you for all the reviews!!

**Comment by marregui** [comment]
retest this please

glitch with SysSnapshotsTest, which I cannot reproduce locally

**Comment by marregui** [comment]
retest this please

next step will be to debug:

```
java.lang.AssertionError: 
Expected: a value equal to or greater than <1593620755390L>
     but: <1593620755320L> was less than <1593620755390L>
	at __randomizedtesting.SeedInfo.seed([4EAC93138A067A77:A85EA0F184C06C92]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at io.crate.integrationtests.SysSnapshotsTest.testQueryAllColumns(SysSnapshotsTest.java:161)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:832)
```
